File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    const group={}
    for(let i=0; i<genres.length; i++){
        const key=genres[i]
        if (!group[key]) {
            group[key] = [];
        }
        group[key].push([i,plays[i]])
    }
    const keys=Object.keys(group);

    keys.sort((a,b)=>{
        const sum1=group[a].reduce((curr,next)=>curr+next[1],0)
        const sum2=group[b].reduce((curr,next)=>curr+next[1],0)
        return sum2-sum1
    })
    for(const key of keys){
        group[key].sort((a,b)=>b[1]-a[1])
        console.log(group[key])
        if(group[key].length===1) answer.push(group[key][0][0])
        else answer.push(group[key][0][0],group[key][1][0])
    }
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> totalPlays = new HashMap<>();
        Map<String, List<Integer>> genreToIndices = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {
            totalPlays.put(genres[i], totalPlays.getOrDefault(genres[i], 0) + plays[i]);
            genreToIndices.computeIfAbsent(genres[i], k -> new ArrayList<>()).add(i);
        }

        // 장르를 총 재생 수 기준으로 정렬
        List<String> sortedGenres = new ArrayList<>(totalPlays.keySet());
        sortedGenres.sort((a, b) -> totalPlays.get(b) - totalPlays.get(a));

        List<Integer> answerList = new ArrayList<>();

        for (String genre : sortedGenres) {
            List<Integer> songs = genreToIndices.get(genre);

            // 곡 인덱스를 재생 수 기준으로 정렬, 같으면 인덱스 오름차순
            songs.sort((a, b) -> {
                if (plays[b] != plays[a]) return plays[b] - plays[a];
                return a - b;
            });

            // 최대 2곡 추가
            for (int i = 0; i < Math.min(2, songs.size()); i++) {
                answerList.add(songs.get(i));
            }
        }

        return answerList.stream().mapToInt(i -> i).toArray();
    }
}
